### PR TITLE
Feature: serve multiple domains on a single instance

### PR DESCRIPTION
**Pull request by igalic**
This Pull Request addresses #94.
We start by adding `custom_domain` to the `blogs` table.

Blogs with custom domains have special routes with higher priority, but administrative routes will remain on the instance host.

Domain administration is left to the user (and/or admin)
(but we might be able to provide a check on whether it's already correctly configured)

### CODE REVIEW

**Comment by igalic** [comment]
after taking a break to work on refactoring with Clippy (#462) , a lot has happened since:

- Blogs can now be edited, and have banners and icons (#460), quite useful, if your front-page is the blog page!
- We have a central Config struct (#494)
- Blogs & Users now have an Fqn field which doesn't need to be recalculated (#457)

Meanwhile, a community member has reported, that Plume works via Tor, without any modifications.

Now, adding the custom domain feature, would possibly require modification.
Regardless, we need to rethink our approach to routing…
…and, i could use some help with that.

Here's some of my own thoughts:

- The route ranking (alone) feels limiting in that regard.
- we have two route functions which do the same thing, only slightly different, so this needs some abstraction
- we have two types of routes: known domain, or default domain (this could would also be taken for unknown domains)
- to properly map onion domains, the `Host` would need to become an array

A closing summary: it would be nice if route functions with `custom_domains: Host(s)` were automatically ranked higher, otherwise, we'd need to split the mounting of routes into: known-domains / unknown or default domain.

File: src/routes/blogs.rs [hunk 1]
```diff
@@ -19,6 +19,38 @@ use plume_models::{
 use routes::{errors::ErrorPage, Page, PlumeRocket};
 use template_utils::Ructe;
 
+#[get("/?<page>", rank = 2)]
+pub fn custom_details(
```
Comment (igalic) [inline src/routes/blogs.rs]: many routes will need duplicated functions, and reordering

in all those cases, we should replace the actual guts of the function with a private one, and all it, once the parameters have been processed successfully

File: plume-models/src/blogs.rs [hunk 2]
```diff
@@ -49,6 +75,7 @@ pub struct Blog {
     pub summary_html: SafeString,
     pub icon_id: Option<i32>,
     pub banner_id: Option<i32>,
+    pub custom_domain: Option<String>,
```
Comment (igalic) [inline plume-models/src/blogs.rs]: btw, i was wondering if this should be an array, and i've come to the conclusion that it should most definitely not be one

File: src/main.rs [hunk 3]
```diff
@@ -137,6 +137,7 @@ Then try to restart Plume
         .mount(
```
[the remark this answers is not in the document]
Comment (fdb-hiroshima) [reply src/main.rs]: as I said (a very long time ago), we could use a [Fairing](https://api.rocket.rs/v0.4/rocket/fairing/trait.Fairing.html#method.on_request) to modify the path from `https://somesite.com/something/idkwhat` to `(https://...)/custom_domain/somesite.com/something/idkwhat`. I think this is the easiest way to do it, the other being to have to reorder each route and add a guard that extract the custom domain of current request, if custom domain there is (like the Host struct, but with an additional check to return forward when it's not a known custom domain)

[the remark this answers is not in the document]
Comment (igalic) [reply src/main.rs]: you did, and i'm still not sure i follow this

let's make a concrete example here:

i have a blog (Words words words) with a custom_domain (https://words.eena.me/) on the instance https://blogs.igalic.co (so, https://blogs.igalic.co/~/words/)

someone makes a request for https://words.eena.me/

what do they see in their browser's URL bar?
what happens in our code?

someone makes a request for https://blogs.igalic.co/~/words/

here, i'm expecting a redirect to https://words.eena.me/

is this what happens, or do we have to do anything special?

thank you very much for your reply, btw, and for your patience

[the remark this answers is not in the document]
Comment (fdb-hiroshima) [reply src/main.rs]: What they see in their url bar : https://words.eena.me/ (unchanged)
What happens in our code : it's treated as `/custom_domain/words.eena.me/` (or anything actually, even `/~/words/` if we want. In a Fairing, we have the host header, the (mutable) request path and access to the database, the sky's the limit)
Someone makes a request for https://blogs.igalic.co/~/words/ , this change nothing. If we want it to redirect to https://words.eena.me/, we have to implement it

Also, made me think that as we use absolute urls, most links could be broken, depending on how the Fairing is made exactly (just concatenating path versus trying to understand them a bit)

**Pull request closed without merging**